datafusion/optimizer/src/unwrap_cast_in_comparison.rs

```diff
@@ -540,10 +602,10 @@ mod tests {
         );
         assert_eq!(optimize_test(expr_lt, &schema), expected);
 
-        // INT32(12) IN (.....)
-        let expr_lt = lit(ScalarValue::Int32(Some(12))).in_list(
+        // cast(INT32(12), INT64) IN (.....)
+        let expr_lt = cast(lit(ScalarValue::Int32(Some(12))), DataType::Int64).in_list(
             vec![
-                lit(ScalarValue::Int32(Some(12))),
+                lit(ScalarValue::Int64(Some(12))),
```

Comment:
Is this change (so that all the `IN` list types are the same) because type coercion will have already been done when this code is now called?

Reply:
Yes.
In this rule, we need to assume that the type coercion has been done for the expr or the plan.